core/src/test/scala/akka/persistence/jdbc/state/JdbcDurableStateSpec.scala

```diff
@@ -0,0 +1,161 @@
+import org.scalatest.{ BeforeAndAfterAll, BeforeAndAfterEach }
+import org.scalatest.concurrent.ScalaFutures
+
+import akka.actor._
+import akka.persistence.PluginSpec
+import akka.persistence.jdbc.db.SlickDatabase
+import akka.persistence.jdbc.config._
+import akka.persistence.jdbc.state.scaladsl.DurableStateStore
+import akka.persistence.jdbc.testkit.internal.{ H2, SchemaType }
+import akka.persistence.jdbc.util.DropCreate
+import akka.serialization.SerializationExtension
+
+abstract class JdbcDurableStateSpec(config: Config, schemaType: SchemaType)
+    extends PluginSpec(config)
+    with AnyWordSpecLike
+    with BeforeAndAfterAll
+    with BeforeAndAfterEach
+    with Matchers
+    with ScalaFutures
+    with DropCreate {
+
+  implicit lazy val ec = system.dispatcher
+
+  val customSerializers = ConfigFactory.parseString("""
+      akka.actor {
+        serializers {
+          my-payload = "akka.persistence.jdbc.state.MyPayloadSerializer"
+        }
+        serialization-bindings {
+          "akka.persistence.jdbc.state.MyPayload" = my-payload
+        }
+      }
+    """)
+
+  lazy val cfg = system.settings.config
+    .getConfig("jdbc-durable-state-store")
+    .withFallback(config.getConfig("jdbc-durable-state-store"))
+    .withFallback(customSerializers.getConfig("akka.actor"))
+
+  lazy val db = SlickDatabase.database(cfg, new SlickConfiguration(cfg.getConfig("slick")), "slick.db")
+  lazy val durableStateConfig = new DurableStateTableConfiguration(cfg)
+  implicit lazy val system: ActorSystem = ActorSystem("JdbcDurableStateSpec", config.withFallback(customSerializers))
+  lazy val serialization = SerializationExtension(system)
+
+  val stateStoreString: DurableStateStore[String]
+  val stateStorePayload: DurableStateStore[MyPayload]
+
+  override def beforeAll(): Unit = {
+    dropAndCreate(schemaType)
+    super.beforeAll()
+  }
+
+  override def afterAll(): Unit = {
+    db.close()
+  }
+
+  "A durable state store" must {
+    "not load a state given an invalid persistenceId" in {
+      whenReady {
+        stateStoreString.getObject("InvalidPersistenceId").failed
+      } { e =>
+        e shouldBe an[java.lang.Exception]
+      }
+    }
+    "add a valid state successfully" in {
+      whenReady {
+        stateStoreString.upsertObject("p123", 1, "a valid string", "t123")
+      } { v =>
+        v shouldBe akka.Done
+      }
+    }
+    "support composite upsert-fetch-repeat loop" in {
+      whenReady {
+        for {
+
+          n <- stateStoreString.upsertObject("p234", 1, "a valid string", "t123")
+          _ = n shouldBe akka.Done
+          g <- stateStoreString.getObject("p234")
+          _ = g.value shouldBe Some("a valid string")
+          u <- stateStoreString.upsertObject("p234", 2, "updated valid string", "t123")
+          _ = u shouldBe akka.Done
+          h <- stateStoreString.getObject("p234")
+
+        } yield h
+      } { v =>
+        v.value shouldBe Some("updated valid string")
+      }
+    }
```

Comment:
add another test when wrong seqNr is used (accidental concurrent writers)

Reply:
You mean 2 writers trying to write with the same sequence number and same persistence id ? What is the expected behavior here ? Note currently the primary key on the table is `persistence_id` ..

Reply:
second write should fail if it doesn't have the expected sequence number

Reply:
added ..